### PR TITLE
Allow all pages if access to robots.txt is forbidden

### PR DESCRIPTION
**Pull request by austin-distilled**

### CODE REVIEW

**Comment by austin-distilled** [comment]
According to [this](https://developers.google.com/webmasters/control-crawl-index/docs/faq?csw=1#h1c) link from Google, if robots.txt is forbidden then "crawlers will generally assume that they can crawl all URLs of the website".

**Comment by austin-distilled** [comment]
I note that the [specification](http://www.robotstxt.org/norobots-rfc.txt) recommends but does not require that a forbidden status means complete restriction. Thus implementing this change would not contradict the specification and might make it more accurate for the real world.

**Comment by b4hand** [comment]
Actually, I think this is not a good idea. We've generally followed the specification and the recommendation is to deny all. Google didn't make the robots.txt specification, and Google gets away with not following it because webmasters are willing to allow Google traffic. Arbitrary crawlers get scrutinized more heavily and are more likely to get banned for even three appearance off misbehavior. Therefore polite crawlers do respect the specification *including* recommendations.

Look at it this way: the default return code for missing pages is 404. 404 for robots.txt already means allow all. Thus if someone has gone to the trouble of changing the response code to 403, they probably no longer mean to allow all.

If you're seeing sites with a 403 robots.txt, I'd recommend that you notify them and tell them that this is an ambiguous signal, which may mean they get omitted from crawls. If that's not what they want then they should change their website.

I'd personally prefer to err on the side of caution when it comes to robots.txt.

File: tests/test_parse.py [hunk 1]
```diff
@@ -352,13 +352,6 @@ def test_skip_line(self):
     ###########################################################################
     # Status code issues
     ###########################################################################
-    def test_status_disallowed(self):
```
Comment (b4hand) [inline tests/test_parse.py]: If we were to accept this change, we'd want an explicit test showing that 403 is allowed instead of just removing the test.

**Comment by TomAnthony** [comment]
Is there scope to add an option for a flag for this? I imagine for a lot of people the use case of this library is similar to ours, which is for fault detection rather than crawling, and in which case you probably have a heavy regard for how Google are handling the situation.

Our system picked up a couple of sites that intermittently returned 403s which triggered a urgent alarm as it showed up as the whole site being blocked, but that is not how Google perceived the situation (as per their docs and what we could observe of their crawling behaviour).

I agree with the point that other crawlers probably get more scrutiny and observing the specification recommendations is probably the best course of action - it is just a shame if the library cannot accommodate what I imagine is the other major use case.

**Comment by b4hand** [comment]
I think a configuration option would be feasible, but the default should definitely be off. Also, be careful of assuming just because Google didn't stop crawling your site that it didn't mean that Bing or other foreign country search engines didn't stop crawling your site. Again, if you were going to treat this as a fault detection, I'd recommend you treat a 403 robots.txt as a strong warning if not a urgent alarm.

**Comment by TomAnthony** [comment]
Yeah - I agree that the option should be off by default.

Agree about treating it as a strong warning - that is our intention. Thanks a lot for the quick replies. :)

**Comment by austin-distilled** [comment]
Hi @b4hand, I'm just revisiting this after other work took priority. I've tried to add the flag in the most lightweight way possible given there was no previous precedent for passing information from RobotsCache to Rules. Let me know if there is an alternative way that you feel is more aligned with the project or if there are any other amendments you'd like to see before merging in. Cheers!

**Comment by b4hand** [comment]
I noticed that you set the executable bit on several of the module files. Those files aren't intended to be executed directly, so I don't think it makes sense to set the executable bit on them.

File: reppy/parser.py [hunk 2]
```diff
@@ -110,9 +110,12 @@ def __init__(self, url, status, content, expires):
...
             self.parse(content)
         elif status in (401, 403):
-            logger.warn('Access disallowed to site %s (%i)' % (
+            logger.warn('Access forbidden to site %s (%i)' % (
                 url, status))
-            self.parse('''User-agent: *\nDisallow: /''')
+            if disallow_forbidden:
+                self.parse('''User-agent: *\nDisallow: /''')
+            else:
+                self.parse('')
```
Comment (b4hand) [inline reppy/parser.py]: This line could probably just be `pass` or better yet just drop the `else` branch entirely.

**Comment by b4hand** [comment]
Otherwise, your approach looks fine to me. I'd like @dlecocq to approve as well, but if you fix those last few things, I think we can probably get this merged.

File: reppy/parser.py (hunk 2 above)
Comment (austin-distilled) [inline reppy/parser.py]: My thinking with this was I was attempting to be consistent with line 122 where self.parse gets called with an empty string. Does this method not need to be called?

**Comment by austin-distilled** [comment]
Thanks @b4hand and @dlecocq. Let me know when you are happy and I'll tidy the commits :+1:

File: reppy/parser.py (hunk 2 above)
Comment (dlecocq) [inline reppy/parser.py]: Off hand, I don't recall why `self.parse('')` is called. My inclination is to leave it as it is and follow the example for the time being.

**Comment by dlecocq** [comment]
LGTM

**Comment by b4hand** [comment]
I'll merge as soon as you collapse the executable bit change back into the original introduction of those bits.

**Comment by austin-distilled** [comment]
I've collapsed all the commits into one. Thanks a lot for your help @b4hand!

**Pull request merged**